### DISCUSSION TITLE
Are the escape sequences for F1, F2 and F4 (documented) wrong?

### DISCUSSION THREAD

**Discussion by plotski** [kovidgoyal/kitty#7650] [Q&A] [answered]
Are the escape sequences for F1, F2 and F4 (documented) wrong?

According to https://sw.kovidgoyal.net/kitty/keyboard-protocol/#functional pressing the keys F1, F2 and F4 should produce the escape sequences `CSI 1 P` (or `CSI 11 ~`), `CSI 1 Q` (or `CSI 12 ~`) and `CSI 1 S` (or `CSI 14 ~`), respectively.

But when I run `kitten show-key -m kitty`, I'm getting `CSI P`, `CSI Q` and `CSI S`.

Same thing for the cursor keys (`CSI 1 A/B/C/D`) and Home/End.

Am I misunderstanding something or are documentation and implementation in disagreement?

And why are there two sequences for some keys? Under which circumstances will the terminal send each one? Is this predictable?

**Comment by kovidgoyal** [comment, marked as answer]
On Sat, Jul 20, 2024 at 04:08:11AM -0700, plotski wrote:
> Are the escape sequences for F1, F2 and F4 (documented) wrong?
>
> According to https://sw.kovidgoyal.net/kitty/keyboard-protocol/#functional pressing the keys F1, F2 and F4 should produce the escape sequences `CSI 1 P` (or `CSI 11 ~`), `CSI 1 Q` (or `CSI 12 ~`) and `CSI 1 S` (or `CSI 14 ~`), respectively.
>
> But when I run `kitten show-key -m kitty`, I'm getting `CSI P`, `CSI Q` and `CSI S`.

1 is the default value, so the two are identical.

You should not be matching bytes directly, but instead parsing the escape code
properly. It takes only about 20 lines of code to write a correct parser
for these sequences.

>
> Same thing for the cursor keys (`CSI 1 A/B/C/D`) and Home/End.
>
> Am I misunderstanding something or are documentation and implementation in disagreement?
>
> And why are there two sequences for some keys? Under which circumstances will the terminal send each one? Is this predictable?

Any given terminal should send just one, but depending on what its legacy
encoding for those keys was it might be one or the other.

In a client library you should recognize both.

**Reply by plotski** [reply]
> 1 is the default value, so the two are identical.

The default value for what, exactly?

**Comment by kovidgoyal** [comment]
On Sat, Jul 20, 2024 at 04:23:36AM -0700, plotski wrote:
> > 1 is the default value, so the two are identical.
>
> The default value for what, exactly?

Quoting the spec:

The second form is used for a few functional keys, such as the Home, End, Arrow keys and F1 … F4, they are enumerated in the Functional key definitions table below. Note that if no modifiers are present the parameters are omitted entirely giving an escape code of the form CSI [ABCDEFHPQS].

**Reply by plotski** [reply]
OK, that's what I thought, but I'm getting `CSI 1 ; 2 P` for Shift+F1,  `CSI 1 ; 3 P` for Alt+F1,  `CSI 1 ; 4 P` for Alt+Shift+F1 and so on. The `1` is always the same. Which keys do I need to press to change the `1`?

Can you point me in the direction of an escape code parser? Sometimes code is easier to understand than English.

**Comment by kovidgoyal** [comment]
Grep for decode_key_event()

**Reply by kovidgoyal** [reply]
And just to answer your question. When there is a trailing letter in the CSI the key code is taken from the letter, therefore, the first number which is normally the key code is ignored, and thus CSI 1 P is the same as CSI P is the same as CSI any single number P

**Reply by plotski** [reply]
OK, I think I got it.

Thank you very much for your great help! Feels like I have a support contract. :)

**Comment by plotski** [comment]
BTW:

https://github.com/kovidgoyal/kitty/blob/ff2baf91225180d335603bd53100aca14f64cfd4/kitty/key_encoding.py#L326

The `R` is not in the documented string: `ABCDEFHPQS`

I believe the `R` was removed due to the cursor position response conflict?

**Comment by kovidgoyal** [comment]
On Sat, Jul 20, 2024 at 06:40:29AM -0700, plotski wrote:
> BTW:
>
> https://github.com/kovidgoyal/kitty/blob/ff2baf91225180d335603bd53100aca14f64cfd4/kitty/key_encoding.py#L326
>
> The `R` is not in the documented string: `ABCDEFHPQS`
>
> I believe the `R` was removed due to the cursor position response conflict?

Yes, it was, but kittens dont use the cursor position report and so I
left it in there, in case the kitten is ever run on an old version of
kitty that did generate R. Modern code should not use it, of course.
